### ISSUE TITLE
possible memory leak reported in resque web

### ISSUE THREAD

**Issue by danmayer** [danmayer/coverband#175]
I got an email with graphs showing constantly growing memory in someone's resque background worker processes after deploying Coverband. I don't have additional details on the version or the config setup. 

I will try to get some additional information but will also see if I can figure something out or reproduce.

**Comment by kbaum** [comment]
Not sure how they have coverband setup with resque but the default background reporting thread wouldn't work as resque forks a new process for each job and the forked process wouldn't have a background thread.

Perhaps we should think about having out of box support for resque?

**Comment by danmayer** [comment]
yeah I am going to ask for more config details thinking we might need to detect resque and do something different.

**Comment by danmayer** [comment]
so it not working makes a lot of sense, but still struggling to figure out why it would leak memory...

**Comment by danmayer** [comment]
tried a bunch of things but can't cause a leak so far... I think we can work on adding out of the box support and perhaps I can get some more info on how things are configured... or while building this feature perhaps a leak will make more sense.

**Comment by danmayer** [comment]
I will investigate this in the next couple of weeks as I get access to a codebase that reproduces this issue.

**Comment by danmayer** [comment]
OK, I have been working with a code base having this issue. I believe I now have a reproducible set of steps, but need to dig in a bit further... I am currently looking at issues / leaks around calling `coverband.start` before the fork and after, which does a bit more work than it needs to. I am thinking the solution will look something like `coverband.restart` since rescue calls `coverband.stop`.

I will likely get all of this verify and pushed up on a branch as the first 4.2 feature.

**Comment by kbaum** [comment]
Interesting.. in the latest coverband, we stop the background thread before the first fork but I think the ruby ::Coverage library will still be `Coverage.running?`. 

https://github.com/danmayer/coverband/blob/master/lib/coverband/integrations/resque.rb#L7

One way to turn coverage off is to just call `Coverage.result` before fork:

```
irb(main):001:0> require 'coverage'
=> true
irb(main):002:0> Coverage.start
=> nil
irb(main):003:0> Coverage.running?
=> true
irb(main):004:0> Coverage.result
=> {}
irb(main):005:0> Coverage.running?
=> false
irb(main):006:0>
```

Curious what the underlying cause of this leak is though.

**Comment by danmayer** [comment]
yeah, we could actually stop Coverage, but I don't think we really want to it is fine to start it and then fork I believe. 

At the moment, I can show memory growth processing jobs if I call Coverband start again in each fork, but haven't fully identified why.

I have also, shown via a test that calling configure multiple times that it will leak redis connection info... but I don't believe that should be happening in the resque flow, but it is good to note.

**Comment by kbaum** [comment]
I think I need to better understand memory growth in forked ruby process. My understanding is that once you fork the process, any memory growth in child is limited to the child. Resque child worker dies after the job executes so any memory growth should go away after the job completes I would think.

**Comment by danmayer** [comment]
I agree this matches my understanding, so I am still a bit confused on the leak. I am still learning a bit more about how systems run and infrastructure so I can do some more in-depth debugging to fully understand the issue.

**Comment by danmayer** [comment]
OK @kbaum I believe I finally figured this out... Some cleanup from all my debugging and additional testing to follow, but this seems to be the root of the Resque issue.

```
#!/usr/bin/env ruby

# This is a small script to illustrate how previous results get wiped out by forking
# this has implications of forking processes like Resque...
# this in the end would cause coverage.array_diff
# with previous results to add NEGATIVE code hits to the stored Coverage
# which in turn causes all sorts of crazy issues.
#
# ruby test/benchmarks/coverage_fork.rb
# in parent before fork
# {"/Users/danmayer/projects/coverband/test/dog.rb"=>[nil, nil, 1, 1, 2, nil, nil]}
# in child after fork
# {"/Users/danmayer/projects/coverband/test/dog.rb"=>[nil, nil, 0, 0, 0, nil, nil]}
# now triggering hits
# {"/Users/danmayer/projects/coverband/test/dog.rb"=>[nil, nil, 0, 0, 3, nil, nil]}
#
# note: that the child now has 0 hits where previously method definitions had 1
require 'coverage'
Coverage.start
load './test/dog.rb'
Dog.new.bark
Dog.new.bark
puts 'in parent before fork'
puts Coverage.peek_result
fork do
  puts 'in child after fork'
  puts Coverage.peek_result
  puts 'now triggering hits'
  Dog.new.bark
  Dog.new.bark
  Dog.new.bark
  puts Coverage.peek_result
end
```

**Comment by kbaum** [comment]
Interesting. I believe I've seen this before when playing with unicorn and forking a while back. It seems like coverage data does not get copied properly on fork. 

Does this explain the memory leak though? 

Previously i had thought this wasn't a big issue since coverband merges the coverage data with what's already stored in redis.

**Comment by danmayer** [comment]
so the merge happens after we try to split the previous_coverage and current_coverage... So the error comes from trying to merge data that is negative.

In terms of if this relates to a memory leak, I am not entirely sure until I run it at volume, I have it running on 2 systems now and I am monitoring...  I can't entirely explain the memory issue but the negative numbers cause some errors and odd behavior so I am hoping it is related.

**Comment by danmayer** [comment]
I updated the code to detect when this situation arises and handle it correctly

**Comment by danmayer** [comment]
so far so good going to move to higher volume systems soon.

**Comment by danmayer** [comment]
OK, on higher volume systems and going well, should be able to officially call this closed after putting on highest volume system early next week.

**Comment by danmayer** [comment]
OK, I am marking this as resolved the 4.2.0.rc1 is live on an extremely high volume site with Resque without any memory issues.
